### PR TITLE
feat: DB-API driver + unit tests

### PR DESCRIPTION
**Pull request by mf2199**
The DB-API driver refactored from [python-spanner-django](https://github.com/q-logic/python-spanner-django) repository, prior to its beta-release.

_Some additional changes have been made to skip unit tests found incompatible with Python 2._

### CODE REVIEW

**Review by c24t — APPROVED** [review]
LGTM, we'll a PR in https://github.com/googleapis/python-spanner-django to remove `spanner_dbapi` and then coordinate releasees. We can follow up with tests for 2.x.

File: google/cloud/spanner_dbapi/cursor.py [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+                # is returned or many, yet mixing results that
+                # are for .fetchone() with those that would result in
+                # many items returns a RuntimeError if .fetchone() is
+                # invoked and vice versa.
+                self._result_set = res
+                # Read the first element so that the StreamedResultSet can
+                # return the metadata after a DQL statement. See issue #155.
+                self._itr = PeekIterator(self._result_set)
+                # Unfortunately, Spanner doesn't seem to send back
+                # information about the number of rows available.
```
Comment (larkee) [inline google/cloud/spanner_dbapi/cursor.py:290]: Yes, unfortunately the row count is sent as part of `ResultSetStats` which is not available until the final `ResultSet` has been received.

See: https://github.com/googleapis/python-spanner/blob/master/google/cloud/spanner_v1/streamed.py#L46

File: google/cloud/spanner_dbapi/parse_utils.py [hunk 2]
```diff
@@ -0,0 +1,546 @@
...
+    """Convert the param to its Cloud Spanner equivalent type.
+
+    :type value: Any
+    :param value: Value to convert to a Cloud Spanner type.
+
+    :rtype: Any
+    :returns: Value converted to a Cloud Spanner type.
+    """
+    if isinstance(value, decimal.Decimal):
+        return float(value)
```
Comment (larkee) [inline google/cloud/spanner_dbapi/parse_utils.py:497]: Converting `Decimal` to `float` can lose precision. Please use `str` instead. 

If you have a good reason for not doing so, you will instead need to clearly document that NUMERICs may lose precision when using this driver.

**Review by larkee — CHANGES_REQUESTED** [review]
LGTM :+1: Just one concern about NUMERIC loss of precision.

File: google/cloud/spanner_dbapi/_helpers.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+
+def _execute_insert_heterogenous(transaction, sql_params_list):
+    for sql, params in sql_params_list:
+        sql, params = sql_pyformat_args_to_spanner(sql, params)
+        param_types = get_param_types(params)
+        res = transaction.execute_sql(sql, params=params, param_types=param_types)
+        # TODO: File a bug with Cloud Spanner and the Python client maintainers
+        # about a lost commit when res isn't read from.
+        _ = list(res)
```
[the remark this answers is not in the document]
Comment (mf2199) [reply google/cloud/spanner_dbapi/_helpers.py:65]: Totally, thanks for the clarification. Done.

File: google/cloud/spanner_dbapi/parse_utils.py (hunk 2 above)
Comment (mf2199) [reply google/cloud/spanner_dbapi/parse_utils.py:497]: Changed as suggested, based on this [recommendation](https://cloud.google.com/spanner/docs/storing-numeric-data#recommendation_store_arbitrary_precision_numbers_as_strings).

**Review by larkee — APPROVED** [review]

**Pull request merged**